### ISSUE TITLE
Thoughts on allowing compression encodings to be prioritize-able?

### ISSUE THREAD

**Issue by tobz** [gotham-rs/gotham#522]
Currently, a user can configure `FileOptions` to support multiple encodings i.e. `gzip` and `br` at the same time.  However, depending on the order of the values in the `Accept-Encoding` header in requests, a suboptimal encoding might be chosen by Gotham when multiple options are available.  For example, if both `gzip` and `br` versions of a file are available, and the `br` version is half the size of the `gzip` version, we might still send the `gzip` version if it was specified earlier in the value list for `Accept-Encoding`.

There are a few approaches, with varying levels of correctness/complexity, that would allow us to shrink or eliminate this problem:

- user-controlled priority of individual encodings
- inline interrogation of files to determine the optimal encoding

User-controlled priority could look something like creating an enumeration/set of constants that let users build their own list -- or use a builder pattern to feed in each option in priority order -- of encoding priorities, which would be coalesced into the enumeration of encodings that arrive via `Accept-Encoding`, ultimately sorting them in the correct order.

Interrogation of the compressed files themselves would obviously be the most optimal, but would require file I/O to occur before actually serving the asset.  Given that file I/O will happen one way or another, it's not clear to me that this is inherently a performance trap or anything, but it is an increase in file I/O operations in order to serve a request.

Both, I believe, are able to be written in such a way that they simply extend the existing `FileOptions` API, although they have differing levels of behind-the-scenes complexity.

So, long story short: does this seem like something that Gotham is willing to deal with? I'm happy to work on a PR to add this functionality, but wanted to get a sense of if this felt like too much complexity being added, or if there was already existing work, etc.

**Comment by msrd0** [comment]
I just had a quick look through the existing code and it seems that we do sort the values of the `Accept-Encoding` header, however, only by the client's priorities. It would probably make sense to have server priorities as well, and those (if present) should probably override the client priorities.

However, I don't think it would make too much sense to check if the server-preferred compression is slightly larger than another compression. This just introduces a small runtime overhead that is unneccessary since this could already be checked after the compression, so e.g. if you preferred brotli over gzip, and the brotli file happens to be larger, just delete the brotli file so the gzip version will be delivered instead (unless your client supported brotli but not gzip but I think that's pretty rare). Unless someone has a use case where this would be different?

**Comment by tobz** [comment]
Practically speaking, you're probably right in like 95% of use cases.  It would just be nice to be right in 100% of cases. :P

There's actually probably more nuance to be had here in terms of interpreting the content of `Accept-Encoding` given that it supports an additional number of settings i.e. quality via `q=...`, etc.  Maybe I'll just take a stab at a handler for the moment rather than trying to bake this all in `gotham` itself right out of the gate. 👍🏻 

Gonna close this ticket for now, appreciate the rubber ducking.
